lib/routing/sfPatternRouting.class.php

```diff
@@ -154,27 +154,31 @@ public function getRoutes()
     return $this->routes;
   }
 
-  /**
-   * @see  sfRouting
-   * @inheritdoc
-   */
-  public function getRoute($name)
-  {
-    if (!array_key_exists($name, $this->routes))
+    /**
+    * @see  sfRouting
+    * @inheritdoc
+     */
+    public function getRoute($name)
     {
-      throw new sfException(sprintf('Route "%s" is not defined.', $name));
-    }
+        if (!array_key_exists($name, $this->routes)) {
+            throw new sfException(sprintf('Route "%s" is not defined.', $name));
+        }
 
-    $route = $this->routes[$name];
+        $route = $this->routes[$name];
 
-    if (is_string($route))
-    {
-      $this->routes[$name] = $route = unserialize($route);
-      $route->setDefaultParameters($this->defaultParameters);
-    }
+        if (is_string($route)) {
+            $decoded = json_decode($route, true);
+            $route   = is_array($decoded) ? sfRoute::jsonUnserialize($decoded) : unserialize($route);
 
-    return $route;
-  }
+            if (is_object($route) && method_exists($route, 'setDefaultParameters')) {
+                $route->setDefaultParameters($this->defaultParameters);
+            }
```

Comment:
It feels you faced an interesting bug, given the presence of `method_exists()`

Reply:
My worry was that the result of deserialisation isn't always an `sfRoute` object, so I added this conditional just to be absolutely sure we don't try and run `setDefaultParameters()` unless the object is of the expected class.